From f3853dbf19636c8f1536ad64899617f1044ff70c Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:17:50 -0400
Subject: [net] netfilter: nf_tables: fix trace of matching non-terminal rule

Message-id: <20141021161750.22888.71910.stgit@localhost.localdomain>
Patchwork-id: 97945
O-Subject: [RHEL7 net PATCH 048/124] netfilter: nf_tables: fix trace of matching non-terminal rule
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 3b084e99a3fabaeb0f9c65a0806cde30f0b2835e

 commit 3b084e99a3fabaeb0f9c65a0806cde30f0b2835e
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Thu May 15 17:18:26 2014 +0200

    netfilter: nf_tables: fix trace of matching non-terminal rule

    Add the corresponding trace if we have a full match in a non-terminal
    rule. Note that the traces will look slightly different than in
    x_tables since the log message after all expressions have been
    evaluated (contrary to x_tables, that emits it before the target
    action). This manifests in two differences in nf_tables wrt. x_tables:

    1) The rule that enables the tracing is included in the trace.

    2) If the rule emits some log message, that is shown before the
       trace log message.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_core.c b/net/netfilter/nf_tables_core.c
index 4e6e681..9bafd17 100644
--- a/net/netfilter/nf_tables_core.c
+++ b/net/netfilter/nf_tables_core.c
@@ -144,8 +144,10 @@ next_rule:
   switch (data[NFT_REG_VERDICT].verdict) {
   case NFT_BREAK:
    data[NFT_REG_VERDICT].verdict = NFT_CONTINUE;
-   /* fall through */
+   continue;
   case NFT_CONTINUE:
+   if (unlikely(pkt->skb->nf_trace))
+    nft_trace_packet(pkt, chain, rulenum, NFT_TRACE_RULE);
    continue;
   }
   break;
-- 
1.7.1